lambdas/build-cri-oauth-request/src/main/java/uk/gov/di/ipv/core/buildcrioauthrequest/helpers/SharedClaimsHelper.java

```diff
@@ -0,0 +1,96 @@
+import uk.gov.di.ipv.core.library.helpers.LogHelper;
+import uk.gov.di.ipv.core.library.verifiablecredential.helpers.VcHelper;
+import uk.gov.di.model.AddressAssertion;
+import uk.gov.di.model.PersonWithDocuments;
+import uk.gov.di.model.PersonWithIdentity;
+
+import java.util.List;
+
+import static uk.gov.di.ipv.core.library.domain.Cri.ADDRESS;
+
+public class SharedClaimsHelper {
+    private static final Logger LOGGER = LogManager.getLogger();
+    public static final String SHARED_CLAIM_ATTR_NAME = "name";
+    public static final String SHARED_CLAIM_ATTR_BIRTH_DATE = "birthDate";
+    public static final String SHARED_CLAIM_ATTR_ADDRESS = "address";
+    public static final String SHARED_CLAIM_ATTR_EMAIL = "emailAddress";
+    public static final String SHARED_CLAIM_ATTR_SOCIAL_SECURITY_RECORD = "socialSecurityRecord";
+
+    private SharedClaimsHelper() {}
+
+    public static SharedClaims generateSharedClaims(
+            String emailAddress, List<VerifiableCredential> vcs, List<String> allowedSharedClaims) {
+        var sharedClaims = new SharedClaims();
+
+        // Email address is provided separately in the session, not from a VC
+        sharedClaims.setEmailAddress(emailAddress);
+
+        // Other shared claims are added from successful VCs
+        vcs.stream()
+                .filter(VcHelper::isSuccessfulVc)
+                .forEach((vc) -> addSharedClaimsFromVc(sharedClaims, vc));
+
+        LOGGER.info(
+                LogHelper.buildLogMessage("Found shared claims")
+                        .with("names", sharedClaims.getName().size())
+                        .with("birthDates", sharedClaims.getBirthDate().size())
+                        .with("addresses", sharedClaims.getAddress().size())
+                        .with("emails", sharedClaims.getEmailAddress() == null ? 0 : 1)
+                        .with(
+                                "socialSecurityRecords",
+                                sharedClaims.getSocialSecurityRecord().size()));
+
+        stripDisallowedSharedClaims(sharedClaims, allowedSharedClaims);
+
+        return sharedClaims;
+    }
+
+    private static void addSharedClaimsFromVc(SharedClaims sharedClaims, VerifiableCredential vc) {
+        var credentialSubject = vc.getCredential().getCredentialSubject();
+
+        if (credentialSubject instanceof PersonWithIdentity personWithIdentity
+                && personWithIdentity.getName() != null) {
+            sharedClaims.getName().addAll(personWithIdentity.getName());
+        }
+        if (credentialSubject instanceof PersonWithIdentity personWithIdentity
+                && personWithIdentity.getBirthDate() != null) {
+            sharedClaims.getBirthDate().addAll(personWithIdentity.getBirthDate());
+        }
+        if (ADDRESS.equals(vc.getCri())
+                && credentialSubject instanceof AddressAssertion addressAssertion
+                && addressAssertion.getAddress() != null) {
```

Comment:
This is a small deviation from the existing implementation - previously it would add addresses from non-address VCs until it found an address VC (at which point it would replace everything with that address).

Given that we only ever expect an address from the address VC, this should be a clearer and easier way of handling it. (And if we need to support addresses on other VCs, it should be relatively clear how to do that).

Reply:
OK - looks like that answers my earlier question